pinot-core/src/main/java/org/apache/pinot/core/transport/DirectOOMHandler.java

```diff
@@ -0,0 +1,100 @@
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.apache.pinot.spi.exception.QueryCancelledException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+// Handling netty direct memory OOM. In this case there is a great chance that multiple channels are receiving
+// large data tables from servers concurrently. We want to close all channels to servers to proactively release
+// the direct memory, because the execution of netty threads can deadlock in allocating direct memory, in which case
+// no one will reach channelRead0.
+public class DirectOOMHandler extends ChannelInboundHandlerAdapter {
+  private static final Logger LOGGER = LoggerFactory.getLogger(DataTableHandler.class);
+  private static final AtomicBoolean DIRECT_OOM_SHUTTING_DOWN = new AtomicBoolean(false);
+  private final QueryRouter _queryRouter;
+  private final ServerRoutingInstance _serverRoutingInstance;
+  private final ConcurrentHashMap<ServerRoutingInstance, ServerChannels.ServerChannel> _serverToChannelMap;
+
+  public void setSilentShutDown() {
+    _silentShutDown = true;
+  }
+
+  private volatile boolean _silentShutDown = false;
+
+  public DirectOOMHandler(QueryRouter queryRouter, ServerRoutingInstance serverRoutingInstance,
+      ConcurrentHashMap<ServerRoutingInstance, ServerChannels.ServerChannel> serverToChannelMap) {
+    _queryRouter = queryRouter;
+    _serverRoutingInstance = serverRoutingInstance;
+    _serverToChannelMap = serverToChannelMap;
+  }
+
+  @Override
+  public void channelInactive(ChannelHandlerContext ctx) {
+    // if we are shutting down channels due to direct memory OOM, we short circuit the channel inactive
+    if (_silentShutDown) {
+      return;
+    }
+    ctx.fireChannelInactive();
+  }
+
+  @Override
+  public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
+    // catch direct memory oom here
+    if (cause instanceof OutOfMemoryError && cause.getMessage().contains("Direct buffer")) {
+      // only one thread can get here and do the shutdown
+      if (DIRECT_OOM_SHUTTING_DOWN.compareAndSet(false, true)) {
+        try {
+          LOGGER.error("Closing ALL channels to servers, as we are running out of direct memory "
+              + "while receiving response from {}", _serverRoutingInstance, cause);
+          // close all channels to servers
+          ctx.channel().close();
+          _serverToChannelMap.keySet().forEach(serverRoutingInstance -> {
+            Channel channel = _serverToChannelMap.get(serverRoutingInstance)._channel;
+            if (channel != null) {
+              DirectOOMHandler directOOMHandler = channel.pipeline().get(DirectOOMHandler.class);
+              if (directOOMHandler != null) {
+                directOOMHandler.setSilentShutDown();
+              }
+              channel.close();
+            }
+            _serverToChannelMap.remove(serverRoutingInstance);
```

Comment:
Wondering if it is necessary to remove the entry for the `serverRoutingInstance` from the `serverToChannelMap` ? 

Won't it be enough to just mark the `value` as NULL since we are closing the channel and the next query will create the channel and populate the map ?

Reply:
On a slightly related note, should we first remove the entry from map / mark value as NULL before the if block from lines 77 to 83 to prevent further use?

Reply:
>Won't it be enough to just mark the value as NULL since we are closing the channel and the next query will create the channel and populate the map ?
No we need to remove the channel in the map, setting it to null will break ServerChannels.sendRequest()

>On a slightly related note, should we first remove the entry from map / mark value as NULL before the if block from lines 77 to 83 to prevent further use?
Yes, swapped.

Reply:
Got it. Thanks